### PR TITLE
NE-1091: Add proxy protocol support to IBMLoadBalancerParameters

### PR DESCRIPTION
**Pull request by SzucsAti**
On IBMCloud users can set a loadbalancer annotation to enable proxy protocol on the loadbalancer: 
`service.kubernetes.io/ibm-load-balancer-cloud-provider-enable-features: "proxy-protocol"`
https://cloud.ibm.com/docs/containers?topic=containers-vpc-lbaas

With this PR users can configure proxy protocol on the ingresscontroller custom resource. Cluster-ingress-operator will read the value of this field and configure the loadbalancer accordingly.
```
  endpointPublishingStrategy:
    loadBalancer:
      providerParameters:
        type: IBM
        ibm:
          protocol: PROXY
      scope: External
    type: LoadBalancerService
```

This PR is based on an other PR which is not yet merged, it implements IBMLoadBalancerParameters:
https://github.com/openshift/api/pull/1208

### CODE REVIEW

File: operator/v1/types_ingress.go [hunk 1]
```diff
@@ -443,7 +443,7 @@ const (
 // +union
 type ProviderLoadBalancerParameters struct {
 	// type is the underlying infrastructure provider for the load balancer.
-	// Allowed values are "AWS", "Azure", "BareMetal", "GCP", "Nutanix",
+	// Allowed values are "AWS", "Azure", "BareMetal", "GCP", "IBM", "Nutanix",
```
Comment (JoelSpeed) [inline operator/v1/types_ingress.go:446]: Where is the change to the discriminator enum to allow this? Can't see a change to `LoadBalancerProviderType` included here

File: operator/v1/types_ingress.go [hunk 2]
```diff
@@ -568,6 +577,39 @@ const (
 	GCPLocalAccess  GCPClientAccess = "Local"
 )
 
+// IBMLoadBalancerParameters provides configuration settings that are
+// specific to IBM Cloud load balancers.
+type IBMLoadBalancerParameters struct {
+	// protocol specifies whether the load balancer uses PROXY protocol to forward connections to
+	// the IngressController. See "service.kubernetes.io/ibm-load-balancer-cloud-provider-enable-features:
+	// "proxy-protocol"" at https://cloud.ibm.com/docs/containers?topic=containers-vpc-lbaas"
+
```
Comment (JoelSpeed) [inline operator/v1/types_ingress.go:586]: Is this blank line expected?

Comment (SzucsAti) [reply operator/v1/types_ingress.go]: Not expected, fixed.

File: operator/v1/types_ingress.go (hunk 1 above)
Comment (SzucsAti) [reply operator/v1/types_ingress.go:446]: It was already changed here: https://github.com/openshift/api/blob/master/operator/v1/types_ingress.go#L473-#L490. Comment was not up to date.

**Comment by JoelSpeed** [comment]
The changes to the API look good here but, given the proximity to branch cuts, I'd like to see a QE approval of the feature before we merge this

/hold

File: operator/v1/types_ingress.go [hunk 3]
```diff
@@ -573,6 +582,39 @@ const (
...
+	// information about PROXY protocol.
+	//
+	// The following values are valid for this field:
+	//
+	// * The empty string.
+	// * "TCP".
+	// * "PROXY".
+	//
+	// The empty string specifies the default, which is TCP without PROXY
+	// protocol.  Note that the default is subject to change.
```
Comment (JoelSpeed) [inline operator/v1/types_ingress.go:611]: Just want to update the wording to be consistent with our standard wording on other APIs
```suggestion
	// Valid values for protocol and TCP, PROXY and omitted.
	// When omitted, this means no opinion and the platform is left to choose a reasonable default, which is subject to change over time.
	// The current default is TCP, without the proxy protocol enabled.
```

File: operator/v1/types_ingress.go [hunk 4]
```diff
@@ -573,6 +582,39 @@ const (
...
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
```
Comment (JoelSpeed) [inline operator/v1/types_ingress.go:614]: You don't need both versions of this, just `+optional` will suffice

File: operator/v1/types_ingress.go (hunk 3 above)
Comment (SzucsAti) [reply operator/v1/types_ingress.go]: Thank you, applied your suggestion.

File: operator/v1/types_ingress.go (hunk 4 above)
Comment (SzucsAti) [reply operator/v1/types_ingress.go]: Removed `// +kubebuilder:validation:Optional`.

**Comment by JoelSpeed** [comment]
API looks good, have you run through any pre-merge validation with the QE team? Could we look to get their ack on this feature?

**Comment by lihongan** [comment]
For pre-merge validation, I think we have no chance to test the api directly. 
Can anyone help to create PR to cluster-ingress-operator with the API bump? then QE can do pre-merge validation with the two PRs. 
cc @SzucsAti @Miciah

**Comment by candita** [comment]
> For pre-merge validation, I think we have no chance to test the api directly. Can anyone help to create PR to cluster-ingress-operator with the API bump? then QE can do pre-merge validation with the two PRs. cc @SzucsAti @Miciah

@SzucsAti can you confirm that https://github.com/openshift/cluster-ingress-operator/pull/812 is a PR that can be used to do pre-merge validation?

**Comment by attiss** [comment]
Hi @lihongan @candita,

Yes, https://github.com/openshift/cluster-ingress-operator/pull/812 is the correct PR for pre-merge validations.

**Comment by lihongan** [comment]
@attiss thanks for confirming. Looks https://github.com/openshift/cluster-ingress-operator/pull/812 needs rebase and some checks were failed.
Anyway, I'll have a try to see if I can build payload with the two PRs.

**Comment by lihongan** [comment]
Oops, cannot build the payload with cluster-bot and it reports:
```
pull request openshift/cluster-ingress-operator#812 needs to be rebased to branch master
```
so the rebase is mandatory

**Comment by SzucsAti** [comment]
Rebased https://github.com/openshift/cluster-ingress-operator/pull/812.

**Comment by lihongan** [comment]
tested with pre-merge validation and passed
```
$ oc get clusterversion
NAME      VERSION                                                   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.13.0-0.ci.test-2023-02-07-013419-ci-ln-k91zc62-latest   True        False         139m    Cluster version is 4.13.0-0.ci.test-2023-02-07-013419-ci-ln-k91zc62-latest

### after updating the ingresscontroller/default and got 
$ oc -n openshift-ingress-operator get ingresscontroller/default -oyaml
<---snip-->
spec:
  endpointPublishingStrategy:
    loadBalancer:
      dnsManagementPolicy: Managed
      providerParameters:
        ibm:
          protocol: PROXY
        type: IBM
      scope: External
    type: LoadBalancerService
<------>
status:
  domain: apps.hongli-ibm.ibmcloud.qe.devcluster.openshift.com
  endpointPublishingStrategy:
    loadBalancer:
      dnsManagementPolicy: Managed
      providerParameters:
        ibm:
          protocol: PROXY
        type: IBM
      scope: External
    type: LoadBalancerService
<------>

### ensure router pods use PROXY protocol
$ oc -n openshift-ingress get deployment/router-default -oyaml
<---snip--->
        - name: ROUTER_USE_PROXY_PROTOCOL
          value: "true"

### ensure the proxy-protocol annotation is added to the LB service 
$ oc -n openshift-ingress get svc/router-default -oyaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    service.kubernetes.io/ibm-load-balancer-cloud-provider-enable-features: proxy-protocol
    service.kubernetes.io/ibm-load-balancer-cloud-provider-ip-type: public
<---snip--->

### ensure routes and ingress operator work well
$ curl https://canary-openshift-ingress-canary.apps.hongli-ibm.ibmcloud.qe.devcluster.openshift.com -k
Healthcheck requested
 
$ oc get co/ingress
NAME      VERSION                                                   AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
ingress   4.13.0-0.ci.test-2023-02-07-013419-ci-ln-k91zc62-latest   True        False         False      153m    
```

File: operator/v1/types_ingress.go (hunk 3 above)
Comment (Miciah) [reply operator/v1/types_ingress.go]: @JoelSpeed, should we update the godoc on the other `protocol` fields likewise?  I can get that in a follow-up if you like.

**Comment by Miciah** [comment]
/lgtm
/hold cancel
@JoelSpeed, are you ready to approve the API change?

File: operator/v1/types_ingress.go (hunk 3 above)
Comment (JoelSpeed) [reply operator/v1/types_ingress.go]: Yeah a follow up would be great if you can, thanks

**Pull request merged**